exasol/python_extension_common/deployment/extract_validator.py

```diff
@@ -0,0 +1,119 @@
+
+    def _create_manifest_udf(self, language_alias: str, udf_name: str):
+        """
+        The SQL statements "ALTER SESSION SET SCRIPT_LANGUAGES" and "ALTER
+        SYSTEM SET SCRIPT_LANGUAGES" doe not check whether the specified
+        BucketFS path exists and has permissions allowing it to be accessed by
+        UDFs.
+
+        Much more a later statement "CREATE SCRIPT" will fail with an error
+        message. Hence we need to use a retry here, as well.
+        """
+        self._pyexasol_conn.execute(
+            f"""
+            CREATE OR REPLACE {language_alias} SET SCRIPT
+                {udf_name}(my_path VARCHAR(256))
+                EMITS (node INTEGER, manifest BOOL) AS
+            import os
+            def run(ctx):
+                ctx.emit(exa.meta.node_id, os.path.isfile(ctx.my_path))
+            /
+            """
+        )
+
+    def _check_all_nodes(self, udf_name: str, nproc: int, manifest: str):
+        result = self._pyexasol_conn.execute(
+            f"""
+            SELECT {udf_name}({manifest})
+            FROM VALUES BETWEEN 1 AND {nproc} t(i) GROUP BY i
+            """
+        ).fetchall()
+        pending = list( x[0] for x in result if not x[1] )
+        self._callback(nproc, pending)
+        if len(pending) > 0:
+            raise ExtractException(
+                f"{len(pending)} of {nproc} nodes are still pending."
+                f" IDs: {pending}")
+
+    def verify_all_nodes(self, schema: str, language_alias: str, bfs_archive_path: bfs.path.PathLike):
+        """
+        Verify if the given bfs_archive_path was extracted on all nodes
+        successfully.
+
+        Raise an ExtractException if the specified bfs_archive_path was not an
+        archive or if after the configured timeout there are still nodes
+        pending, for which the extraction could not be verified, yet.
+        """
+        manifest = f"{bfs_archive_path.as_udf_path()}/{MANIFEST_FILE}"
+        if manifest is None:
+            raise ExtractException(
+                f"{bfs_archive_path} does not point to an archive"
+                f" which could contain a file {MANIFEST_FILE}")
+        nproc = self._pyexasol_conn.execute("SELECT nproc()").fetchone()
+        udf_name = _udf_name(schema)
+        start = datetime.now()
+        try:
+            for attempt in Retrying(
+                    wait=wait_fixed(self._interval),
+                    stop=stop_after_delay(self._timeout),
+                    reraise=True):
+                with attempt:
+                    self._create_manifest_udf(language_alias, udf_name)
+            elapsed = datetime.now() - start
+            remaining = self._timeout - elapsed
+            for attempt in Retrying(
+                    wait=wait_fixed(self._interval),
+                    stop=stop_after_delay(remaining),
+                    reraise=True):
+                with attempt:
+                    self._check_all_nodes(udf_name, nproc, manifest)
```

Comment:
would maybe extract that into a method

Reply:
I intentionally switched from a method annotated with `@retry()` as in this case the timeout needs to be calculated dynamically. If moving this to a separate method, the actual retry still needs to be applied to submethod or a `with` context.

Please confirm, if you still prefer having a separate method, calling another method with retry, that then finally calls `self._check_all_nodes()`.

Reply:
See above

Reply:
see above

```
            elapsed = datetime.now() - start
            remaining = self._timeout - elapsed
            for attempt in Retrying(
                    wait=wait_fixed(self._interval),
                    stop=stop_after_delay(remaining),
                    reraise=True):
                with attempt:
                    self._check_all_nodes(udf_name, nproc, manifest)
```

Reply:
extracted into separate method. See latest push.

Reply:
extracted into separate method. See latest push.